src/ArchiveFolder.cs

```diff
@@ -33,15 +33,24 @@ public async Task<IChildFolder> CreateFolderAsync(string name, bool overwrite =
 
         var id = Id + name + ZIP_DIRECTORY_SEPARATOR;
         var key = GetKey(id);
-
         var subfolders = await GetSubfoldersAsync(cancellationToken);
-        if (subfolders.TryGetValue(key, out var folder))
-        {
-            if (!overwrite)
-                return folder;
-        }
-
-        await RemoveSubfolder(key, cancellationToken);
+        
+        // Folder doesn't already exist, simply create it
+        if (!subfolders.TryGetValue(key, out var folder))
+            return await AddSubfolder(key, name, cancellationToken);
+        
+        // Folder already exists and caller doesn't want to overwrite it
+        if (!overwrite)
+            return folder;
+            
+        // Folder already exists and caller wants to overwrite it,
+        // so get the parent and attempt to delete the existing
+        // one before creating a new one.
+        var parent = await folder.GetParentAsync(cancellationToken);
+        if (parent is not IModifiableFolder modifiableParent)
+            throw new IOException($"A folder with the name '{name}' already exists in parent folder " +
+                                  $"'{parent?.Id ?? null}' and is not modifiable.");
+        await modifiableParent.DeleteAsync(folder, cancellationToken);
```

Comment:
This is the real change in this method: actually call `DeleteAsync` to get the full recursive removal of the folder *before* creating it again.